src/iterate.jl

```diff
@@ -0,0 +1,149 @@
+    # One ExtrapSpec per Axis
+    map(KnotIterator, k, bc)
+end
+function KnotIterator(k::NTuple{N,AbstractArray}, bc::BoundaryCondition) where {N}
+    # All Axes use the same BoundaryCondition
+    map(x -> KnotIterator(x, bc), k)
+end
+function KnotIterator(k::AbstractArray{T}, bc::ET) where {T,ET <: ExtrapDimSpec}
+    # Prior dispatches will end up here: Axis knots: k and boundary condition bc
+    KnotIterator{T,ET}(k, bc)
+end
+
+const RepeatKnots = Union{Periodic,Reflect}
+IteratorSize(::Type{KnotIterator}) = SizeUnknown()
+IteratorSize(::Type{KnotIterator{T}}) where {T} = SizeUnknown()
+
+IteratorSize(::Type{KnotIterator{T,ET}}) where {T,ET} = _knot_iter_size(ET)
+_knot_iter_size(::Type{<:BoundaryCondition}) = HasLength()
+_knot_iter_size(::Type{<:RepeatKnots}) = IsInfinite()
+_knot_iter_size(::Type{Tuple{RevBC, FwdBC}}) where {RevBC,FwdBC} = _knot_iter_size(FwdBC)
+
+length(iter::KnotIterator) = _knot_length(iter, IteratorSize(iter))
+_knot_length(iter::KnotIterator, ::HasLength) = iter.nknots
+size(iter::KnotIterator) = (length(iter),)
+
+IteratorEltype(::Type{KnotIterator{T,ET}}) where {T,ET} = HasEltype()
+eltype(::Type{KnotIterator{T,ET}}) where {T,ET} = T
+
+"""
+    knots(itp::AbstractInterpolation)
+    knots(etp::AbstractExtrapolation)
+
+Returns an iterator over knot locations for an AbstractInterpolation or
+AbstractExtrapolation.
+
+Iterator will yeild scalar values for interpolations over a single dimension,
+and tuples of coordinates for higher dimension interpolations. Iteration over
+higher dimension is taken as the product of knots on each dimensions.
+
+ie. Iterator.product(knots on first dim, knots on 2nd dim,...)
+
+Extrapolations with Periodic or Reflect boundary conditions, will produce an
+infinite sequence of knots.
+
+# Example
+```julia-repl
+julia> etp = LinearInterpolation([1.0, 1.2, 2.3, 3.0], rand(4); extrapolation_bc=Periodic())
+julia> Iterators.take(knots(etp), 5) |> collect
+5-element Array{Float64,1}:
+ 1.0
+ 1.2
+ 2.3
+ 3.0
+ 3.2
+```
+"""
+function knots(itp::AbstractInterpolation)
+    # Constuct seperate KnotIterator for each dimension, and combine them using
+    # Iterators.product
+    k = getknots(itp)
+    bc = Throw()
+    iter = KnotIterator(k, bc)
+    length(iter) == 1 ? iter[1] : Iterators.product(iter...)
+end
+
+function knots(etp::AbstractExtrapolation)
+    # Constuct seperate KnotIterator for each dimension, and combine them using
+    # Iterators.product
+    k = getknots(etp)
+    bc = etpflag(etp)
+    iter = KnotIterator(k, bc)
+    length(iter) == 1 ? iter[1] : Iterators.product(iter...)
+end
+
+# For non-repeating ET's iterate through once
+iterate(iter::KnotIterator) where {T} = iterate(iter, 1)
+iterate(iter::KnotIterator, idx::Integer) = idx <= iter.nknots ? (iter.knots[idx], idx+1) : nothing
+
+# For repeating knots state is the knot index + offset value
+function iterate(iter::KnotIterator{T,ET}) where {T,ET <: @FwdExtrapSpec RepeatKnots}
```

Comment:
Codecov is saying this the only line left not covered by tests. I'm not sure why though.